File: verenigingen/verenigingen/report/anbi_periodic_agreements/anbi_periodic_agreements.js
/**
 * @fileoverview ANBI Periodic Agreements Report - JavaScript Configuration
 *
 * This file manages periodic donation agreements for ANBI (Algemeen Nut Beogende Instelling)
 * tax compliance, supporting Dutch charitable organization tax benefits and donor management.
 *
 * BUSINESS PURPOSE:
 * Manages long-term donation commitments with Dutch tax implications:
 * - Track multi-year periodic donation agreements for ANBI compliance
 * - Monitor agreement status, expiration, and completion rates
 * - Support donor stewardship through automated renewal processes
 * - Generate tax receipts for charitable deduction purposes
 * - Ensure compliance with Dutch charitable organization regulations
 * - Facilitate donor retention through proactive engagement
 *
 * ANBI COMPLIANCE FEATURES:
 * - ANBI eligibility tracking and filtering
 * - Tax receipt generation with appropriate documentation
 * - Periodic agreement monitoring for regulatory compliance
 * - Automated renewal reminder system for donor retention
 * - Export capabilities for tax authority reporting
 *
 * AGREEMENT LIFECYCLE MANAGEMENT:
 * - Status tracking (Active, Completed, Cancelled, Expired)
 * - Expiration monitoring with proactive notifications
 * - Completion percentage tracking for donor engagement
 * - Payment frequency management (Monthly, Quarterly, Annually)
 * - Minimum threshold filtering for significant donations
 *
 * VISUAL INDICATORS:
 * - Green: Active agreements and ANBI-eligible donations
 * - Orange: Expiring soon (requires immediate attention)
 * - Red: Cancelled agreements or critical deadlines (≤30 days)
 * - Blue: Completed agreements (successful fulfillment)
 * - Gray: Expired agreements (historical record)
 *
 * AUTOMATED OPERATIONS:
 * - Renewal reminder system with configurable timing
 * - Bulk tax receipt generation for reporting periods
 * - Agreement export for external analysis and compliance
 * - Donor communication automation for stewardship
 *
 * INTEGRATION ARCHITECTURE:
 * - Links to Donor DocType for comprehensive donor profiles
 * - Connects with Periodic Donation Agreement workflows
 * - Integrates with ANBI operations for tax compliance
 * - Supports donor communication and stewardship systems
 *
 * @author Frappe Technologies Pvt. Ltd.
 * @since 2025
 * @category Donor Management / Tax Compliance
 * @requires frappe.query_reports
 * @requires verenigingen.api.periodic_donation_operations
 * @requires verenigingen.api.anbi_operations
 */

// Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.query_reports['ANBI Periodic Agreements'] = {
	filters: [
		{
			fieldname: 'status',
			label: __('Status'),
			fieldtype: 'Select',
			options: '\nActive\nCompleted\nCancelled\nExpired'
		},
		{
			fieldname: 'donor',
			label: __('Donor'),
			fieldtype: 'Link',
			options: 'Donor'
		},
		{
			fieldname: 'anbi_eligible',
			label: __('ANBI Eligible'),
			fieldtype: 'Select',
			options: '\nYes\nNo',
			description: __('Filter by ANBI eligibility status')
		},
		{
			fieldname: 'expiring_in_days',
			label: __('Expiring in Days'),
			fieldtype: 'Int',
			description: __('Show agreements expiring within specified days')
		},
		{
			fieldname: 'payment_frequency',
			label: __('Payment Frequency'),
			fieldtype: 'Select',
			options: '\nMonthly\nQuarterly\nAnnually'
		},
		{
			fieldname: 'min_annual_amount',
			label: __('Minimum Annual Amount'),
			fieldtype: 'Currency',
			description: __('Show only agreements with annual amount >= this value')
		}
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === 'status') {
			if (value.includes('Active')) {
				if (value.includes('Expiring Soon')) {
					value = `<span class="indicator-pill orange">${value}</span>`;
				} else {
					value = `<span class="indicator-pill green">${value}</span>`;
				}
			} else if (value === 'Completed') {
				value = `<span class="indicator-pill blue">${value}</span>`;
			} else if (value === 'Cancelled') {
				value = `<span class="indicator-pill red">${value}</span>`;
			} else if (value === 'Expired') {
				value = `<span class="indicator-pill gray">${value}</span>`;
			}
		}

		if (column.fieldname === 'commitment_type') {
			if (value.includes('ANBI')) {
				value = `<span style="color: green; font-weight: bold;">${value}</span>`;
			} else if (value.includes('Pledge')) {
				value = `<span style="color: blue;">${value}</span>`;
			}
		}

		if (column.fieldname === 'anbi_eligible' && value) {
			value = `<span class="indicator-pill green">${__('Yes')}</span>`;
		} else if (column.fieldname === 'anbi_eligible') {
			value = `<span class="indicator-pill gray">${__('No')}</span>`;
		}

		if (column.fieldname === 'completion_percentage') {
			let color = 'red';
			if (value >= 80) { color = 'green'; } else if (value >= 50) { color = 'orange'; }
			value = `<span style="color: ${color}; font-weight: bold;">${value}%</span>`;
		}

		if (column.fieldname === 'days_remaining') {
			if (value > 0 && value <= 30) {
				value = `<span style="color: red; font-weight: bold;">${value}</span>`;
			} else if (value > 30 && value <= 90) {
				value = `<span style="color: orange; font-weight: bold;">${value}</span>`;
			}
		}

		return value;
	},

	onload(report) {
		// Add custom buttons
		report.page.add_inner_button(__('Send Renewal Reminders'), () => {
			frappe.prompt({
				fieldname: 'days',
				label: __('Days Before Expiry'),
				fieldtype: 'Int',
				default: 90,
				description: __('Send reminders to agreements expiring within these many days')
			}, (values) => {
				frappe.call({
					method: 'verenigingen.api.periodic_donation_operations.send_renewal_reminders',
					args: {
						days_before_expiry: values.days
					},
					callback(r) {
						if (r.message) {
							frappe.msgprint(__('{0} renewal reminder emails sent', [r.message.sent_count]));
						}
					}
				});
			}, __('Send Renewal Reminders'));
		});

		report.page.add_inner_button(__('Generate Tax Receipts'), () => {
			const filters = report.get_filter_values();
			filters.generate_receipts = true;

			frappe.call({
				method: 'verenigingen.api.anbi_operations.generate_tax_receipts',
				args: {
					filters
				},
				callback(r) {
					if (r.message) {
						frappe.msgprint(__('{0} tax receipts generated', [r.message.generated_count]));
					}
				}
			});
		});

		report.page.add_inner_button(__('Export Agreements'), () => {
			frappe.call({
				method: 'verenigingen.api.periodic_donation_operations.export_agreements',
				args: {
					filters: report.get_filter_values()
				},
				callback(r) {
					if (r.message && r.message.file_url) {
						window.open(r.message.file_url);
					}
				}
			});
		});
	}
};

File: verenigingen/verenigingen/report/anbi_periodic_agreements/anbi_periodic_agreements.py
# Copyright (c) 2025, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import add_days, date_diff, flt, formatdate, getdate, today


def execute(filters=None):
    if not filters:
        filters = {}

    # Check if ANBI functionality is enabled
    anbi_enabled = frappe.db.get_single_value("Verenigingen Settings", "enable_anbi_functionality")
    if not anbi_enabled:
        frappe.msgprint(_("ANBI functionality is not enabled. Please enable it in Verenigingen Settings."))
        return [], []

    columns = get_columns()
    data = get_data(filters)

    return columns, data


def get_columns():
    return [
        {
            "label": _("Agreement #"),
            "fieldname": "agreement_number",
            "fieldtype": "Link",
            "options": "Periodic Donation Agreement",
            "width": 140,
        },
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 90},
        {"label": _("Type"), "fieldname": "commitment_type", "fieldtype": "Data", "width": 200},
        {"label": _("Donor"), "fieldname": "donor", "fieldtype": "Link", "options": "Donor", "width": 150},
        {"label": _("Donor Name"), "fieldname": "donor_name", "fieldtype": "Data", "width": 200},
        {"label": _("Duration"), "fieldname": "duration", "fieldtype": "Data", "width": 100},
        {"label": _("Start Date"), "fieldname": "start_date", "fieldtype": "Date", "width": 100},
        {"label": _("End Date"), "fieldname": "end_date", "fieldtype": "Date", "width": 100},
        {"label": _("Days Left"), "fieldname": "days_remaining", "fieldtype": "Int", "width": 80},
        {"label": _("Annual Amount"), "fieldname": "annual_amount", "fieldtype": "Currency", "width": 120},
        {
            "label": _("Payment Frequency"),
            "fieldname": "payment_frequency",
            "fieldtype": "Data",
            "width": 120,
        },
        {"label": _("Total Donated"), "fieldname": "total_donated", "fieldtype": "Currency", "width": 120},
        {
            "label": _("Completion %"),
            "fieldname": "completion_percentage",
            "fieldtype": "Percent",
            "width": 100,
        },
        {"label": _("Expected Total"), "fieldname": "expected_total", "fieldtype": "Currency", "width": 120},
        {
            "label": _("Next Expected"),
            "fieldname": "next_expected_donation",
            "fieldtype": "Date",
            "width": 110,
        },
        {"label": _("ANBI Eligible"), "fieldname": "anbi_eligible", "fieldtype": "Check", "width": 100},
    ]


def get_data(filters):
    conditions = get_conditions(filters)

    # Get agreement data
    agreement_data = frappe.db.sql(
        """
        SELECT
            pda.name,
            pda.agreement_number,
            pda.status,
            pda.commitment_type,
            pda.donor,
            pda.donor_name,
            pda.agreement_duration_years,
            pda.start_date,
            pda.end_date,
            pda.annual_amount,
            pda.payment_frequency,
            pda.payment_amount,
            pda.total_donated,
            pda.donations_count,
            pda.last_donation_date,
            pda.next_expected_donation,
            pda.anbi_eligible,
            pda.docstatus
        FROM `tabPeriodic Donation Agreement` pda
        WHERE pda.docstatus = 1
        {conditions}
        ORDER BY pda.end_date ASC, pda.agreement_number ASC
    """.format(
            conditions=conditions
        ),
        filters,
        as_dict=1,
    )

    data = []
    today_date = getdate(today())

    for row in agreement_data:
        # Calculate days remaining
        days_remaining = 0
        if row.end_date and row.status == "Active":
            days_remaining = date_diff(row.end_date, today_date)
            if days_remaining < 0:
                days_remaining = 0

        # Calculate expected total based on duration
        expected_total = 0
        if row.start_date and row.end_date and row.annual_amount:
            years = date_diff(row.end_date, row.start_date) / 365.25
            expected_total = flt(row.annual_amount * years, 2)

        # Calculate completion percentage
        completion_percentage = 0
        if expected_total > 0:
            completion_percentage = flt((row.total_donated or 0) / expected_total * 100, 2)

        # Format duration
        duration = row.agreement_duration_years or ""
        if duration:
            # Extract years from string like "5 Years (ANBI Minimum)"
            try:
                years = int(duration.split()[0])
                duration = f"{years} Year{'s' if years > 1 else ''}"
            except:
                pass

        # Add status indicators
        status = row.status
        if row.status == "Active" and days_remaining > 0 and days_remaining <= 90:
            status = f"{row.status} (Expiring Soon)"
        elif row.status == "Active" and days_remaining == 0:
            status = "Expired"

        data.append(
            {
                "agreement_number": row.agreement_number,
                "status": status,
                "commitment_type": row.commitment_type
                or ("ANBI Agreement" if row.anbi_eligible else "Donation Pledge"),
                "donor": row.donor,
                "donor_name": row.donor_name,
                "duration": duration,
                "start_date": row.start_date,
                "end_date": row.end_date,
                "days_remaining": days_remaining if row.status == "Active" else 0,
                "annual_amount": row.annual_amount,
                "payment_frequency": row.payment_frequency,
                "total_donated": row.total_donated or 0,
                "completion_percentage": completion_percentage,
                "expected_total": expected_total,
                "next_expected_donation": row.next_expected_donation if row.status == "Active" else None,
                "anbi_eligible": row.anbi_eligible,
            }
        )

    return data


def get_conditions(filters):
    conditions = []

    if filters.get("status"):
        conditions.append("pda.status = %(status)s")

    if filters.get("donor"):
        conditions.append("pda.donor = %(donor)s")

    if filters.get("anbi_eligible") == "Yes":
        conditions.append("pda.anbi_eligible = 1")
    elif filters.get("anbi_eligible") == "No":
        conditions.append("(pda.anbi_eligible = 0 OR pda.anbi_eligible IS NULL)")

    if filters.get("expiring_in_days"):
        future_date = add_days(today(), int(filters.get("expiring_in_days")))
        conditions.append(
            f"pda.end_date <= '{future_date}' AND pda.end_date >= '{today()}' AND pda.status = 'Active'"
        )

    if filters.get("payment_frequency"):
        conditions.append("pda.payment_frequency = %(payment_frequency)s")

    if filters.get("min_annual_amount"):
        conditions.append("pda.annual_amount >= %(min_annual_amount)s")

    return " AND " + " AND ".join(conditions) if conditions else ""
